### DISCUSSION TITLE
Can't get the ical4j bundle to activate in felix 6.0.3

### DISCUSSION THREAD

**Discussion by rudivk** [ical4j/ical4j#555] [General]
Hi all,

I must be missing a dependency somewhere but I cannot figure out which one. I have copied the ical4j-3.2.1.jar into the felix bundle directory and the only missing requirement is:

org.osgi.framework.BundleException: Unable to resolve ical4j [45](R 45.0): missing requirement [ical4j [45](R 45.0)] osgi.extender; (osgi.extender=osgi.serviceloader.processor) Unresolved requirements: [[ical4j [45](R 45.0)] osgi.extender; (osgi.extender=osgi.serviceloader.processor)]

which I believe is provided by https://aries.apache.org/documentation/modules/spi-fly.html
The manifest file of that component clearly shows that it provides the capability osgi.serviceloader.processor:
Provide-Capability: osgi.extender;osgi.extender="osgi.serviceloader.re
 gistrar";version:Version="1.0",osgi.extender;osgi.extender="osgi.serv
 iceloader.processor";version:Version="1.0";uses:="org.apache.aries.sp
 ifly"

I opted for the dynamic weaving framework extension (https://aries.apache.org/documentation/modules/spi-fly.html#_dynamic_weaving_framework_extension) as that is extending the system.bundle and therefore is always active long before any "custom" bundles from the application directory are loaded/activated.

I also modified felix/conf/config.properties and added:

org.apache.aries.spifly.auto.providers=" \
    net.fortuna.ical4j.*;vendor=ical4j

I am not 100% sure if net.fortuna.ical4j.* is the correct syntax because the ical4j bundle says that it's symbolic name is ‘ical4j” . However, I also tried with:
org.apache.aries.spifly.auto.providers=" \
    ical4j.*;vendor=ical4j
and that did not help either.

So with all the above, I believe that the osgi.serviceloader.processor dependency should be covered but the bundle does not activate. What am I doing wrong? How are you typically deploying the ical4j jar in a felix environment?

**Comment by benfortuna** [comment]
So these are the bundles I have used to activate ical4j v4  in Felix, but I think it should be the same for v3.2.1 (asm-all may be the crucial missing bundle in your case):

* org.codehaus.groovy:groovy:2.5.4
* org.apache.aries.spifly:org.apache.aries.spifly.dynamic.bundle:1.2.3
* org.ow2.asm:asm-all:5.2
* org.mnode.ical4j:ical4j:4.0.0-alpha2
* org.threeten:threeten-extra:1.5.0
* javax.cache:cache-api:1.1.1
* org.ehcache:ehcache:3.4.0

**Reply by rudivk** [reply]
Ok, I upgraded to 4.0.0-alpha11 and had to add commons-codec-1.11.jar and commons-lang3-3.9.jar but that's fine.
I seem to have another issue though with parsing content and extracting the start and end. I will create an issue for that as I believe the 3.1.2 parsed it just fine.

**Comment by rudivk** [comment]
Yesterday, the code finally made it to the testing team and despite the fact that the ical4j bundle loads and activates just fine in felix and that my "mailserver" bundle is pulling in the ical4j bundle (see screenshot), I am getting a ClassNotFoundExceptions:

`java.lang.NoClassDefFoundError: Could not initialize class net.fortuna.ical4j.data.CalendarParserFactory
        at net.fortuna.ical4j.data.CalendarBuilder.<init>(CalendarBuilder.java:77)
        at com.linkedstrategies.infinity.mailserver.impl.CalendarEventImpl.<init>(CalendarEventImpl.java:35)
        at com.linkedstrategies.infinity.mailserver.impl.MailImpl.collectCalendarEvents(MailImpl.java:303)
        at com.linkedstrategies.infinity.mailserver.impl.MailImpl.collectCalendarEvents(MailImpl.java:305)
        at com.linkedstrategies.infinity.mailserver.impl.MailImpl.getCalendarEvents(MailImpl.java:284)
        at com.linkedstrategies.infinity.campaigns.impl.MailReceivedMessageProcessor.toCalendarEventsWithFilePath(MailReceivedMessageProcessor.java:307)
        at com.linkedstrategies.infinity.campaigns.impl.MailReceivedMessageProcessor.process(MailReceivedMessageProcessor.java:192)
        at com.linkedstrategies.infinity.campaigns.impl.MailReceivedMessageProcessor.process(MailReceivedMessageProcessor.java:139)
        at com.linkedstrategies.infinity.infrastructure.jms.impl.ConsumerThread.doProcess(ConsumerThread.java:164)
        at com.linkedstrategies.infinity.infrastructure.jms.impl.ConsumerThread.runWhileActive(ConsumerThread.java:106)
        at com.linkedstrategies.infinity.infrastructure.jms.impl.ConsumerThread.run(ConsumerThread.java:70)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)
`

<img width="1710" alt="Screenshot 2022-04-28 at 08 32 51" src="https://user-images.githubusercontent.com/16107266/165691496-3f44d553-9ac6-4167-a541-1af463cf87fd.png">

**Comment by benfortuna** [comment]
Ok, I don't think I've tried the parser in OSGi yet, so it may be a bug. I'll try to recreate this as an integration test and let you know.

**Reply by rudivk** [reply]
I appreciate that you maintain this library in your spare time but we are looking to build a proof of concept around the library. Can you give us an estimate time to fix?

**Comment by benfortuna** [comment]
Sure, I started looking into adding OSGi integration tests, but I think that will take quite a bit longer to resolve. In the meantime I'll see if I can manually debug in an OSGi environment, will give some feedback on this in the next few days.

**Comment by benfortuna** [comment]
So I've done some testing on the ical4j 3.2.2 release using Apache Karaf, and have made some changes to fix issues with activation.

First, I've added some more optional resolution for packages and features that aren't necessarily required for basic icalendar parsing. 

This allowed me to load ical4j and the new [ical4j-validator](https://github.com/ical4j/ical4j-validator) project and I was able to successfully parse a file.

I had to add the following Karaf features and bundles to get it to work:

* `feature:install webconsole`
* `feature:install felix-http` (should auto install servlet, http, and http-whiteboard)
* `feature:install scr`
* Dependency bundles: `org.apache.commons.commons-codec`, `org.apache.commons.lang3`, `com.sun.activation.jakarta.activation`, `com.sun.mail.jakarta.mail`
* iCal4j bundles: `ical4j`, `ical4j-validator`

~~I haven't released the changes yet, but will do it shortly (look for ical4j 3.2.3)~~
ical4j 3.2.3 is now released.

**Comment by rudivk** [comment]
Dunno how but I kinda waited for a confirmation of the 2.3.2 release to be published. I checked today and it was there, I integrated it and downloaded the new dependencies into our bundle directory and the class loading issue is definitely gone. You got that fixed alright, great job and thank you for that!

However (there is always one like this right), the first calendar invite I tried to parse is producing a parse exception. Can I pass it to you here or do you want me to create a separate ticket?

**Reply by benfortuna** [reply]
Sure you can paste the error here. I also got a parse exception when I first tried, so I usually recommend to enable the following compatibility hints (e.g. in `ical4j.properties`):

```
ical4j.unfolding.relaxed=true
ical4j.parsing.relaxed=true
```

The thing is, due to the way OSGi class loaders work, the compatibility hints don't always work properly. So for now I am just explicitly enabling relaxed unfolding with an `UnfoldingReader` instance (see [here](https://github.com/ical4j/ical4j-validator/blob/4a59f0b68ece7767f62efc57de5e1b662f843c80/src/main/java/org/ical4j/validator/ICalendarValidatorServlet.java#L48) for details).

**Comment by rudivk** [comment]
Ok, I will try that by modifying the start script, adding the options using -Dical4j.unfolding.relaxed=true -Dical4j.parsing.relaxed=true.
I also had -Dnet.fortuna.ical4j.timezone.cache.impl=net.fortuna.ical4j.util.MapTimeZoneCache in there, do you think I still need that or does it confuse matters?

I have also attached the stacktrace and the SMTP enveloppe that causes it.

[ls-calendar-parse-exception.txt](https://github.com/ical4j/ical4j/files/8808501/ls-calendar-parse-exception.txt)
[95a274c2-4c57-4196-b5fc-fe698221ba13.txt](https://github.com/ical4j/ical4j/files/8808502/95a274c2-4c57-4196-b5fc-fe698221ba13.txt)
.

**Reply by benfortuna** [reply]
Yes, I suspect it is an issue with unfolding, as it appears to be trying to parse a wrapped line as separate content. If the command line compatibility hints don't work try using an explicit relaxed `UnfoldingReader` as mentioned in my previous comment.
